src/main/java/org/elasticsearch/repositories/gce/GoogleCloudStorageRepository.java

```diff
@@ -0,0 +1,139 @@
+        super(name.name(), repositorySettings, indexShardRepository);
+
+        String projectId = repositorySettings.settings().get(PROJECT_ID, componentSettings.get(PROJECT_ID, settings.get(REPOSITORIES_GS + PROJECT_ID, settings.get("cloud.gce." + PROJECT_ID))));
+        if (!Strings.hasText(projectId)) {
+            throw new RepositoryException(name.name(), "No project id defined for Google Cloud Storage repository");
+        }
+
+        String bucketName = repositorySettings.settings().get(BUCKET, componentSettings.get(BUCKET, settings.get(REPOSITORIES_GS + BUCKET)));
+        if (!Strings.hasText(bucketName)) {
+            throw new RepositoryException(name.name(), "No bucket defined for Google Cloud Storage repository");
+        }
+
+        String bucketLocation = repositorySettings.settings().get(BUCKET_LOCATION, componentSettings.get(BUCKET_LOCATION));
+        if (bucketLocation == null) {
+            // Checks for a default location
+            String defaultLocation = repositorySettings.settings().get(REPOSITORIES_GS + BUCKET_LOCATION, settings.get(REPOSITORIES_GS + BUCKET_LOCATION));
+            if (defaultLocation != null) {
+                defaultLocation = defaultLocation.toUpperCase(Locale.ENGLISH);
+                if ("US".equals(defaultLocation)) {
+                    // Default location - setting location to null
+                    bucketLocation = null;
+                } else if ("EU".equals(defaultLocation)) {
+                    bucketLocation = "EU";
+                } else if ("ASIA".equals(defaultLocation)) {
+                    bucketLocation = "ASIA";
+                }
+            }
+        }
+
+        this.chunkSize = repositorySettings.settings().getAsBytesSize(CHUNK_SIZE, componentSettings.getAsBytesSize(CHUNK_SIZE, new ByteSizeValue(100, ByteSizeUnit.MB)));
+        this.compress = repositorySettings.settings().getAsBoolean(COMPRESS, componentSettings.getAsBoolean(COMPRESS, false));
+
+        String basePath = repositorySettings.settings().get(BASE_PATH, null);
+        if (Strings.hasLength(basePath)) {
+            // Remove starting / if any
+            basePath = Strings.trimLeadingCharacter(basePath, '/');
+            BlobPath path = new BlobPath();
+            for (String elem : Strings.splitStringToArray(basePath, '/')) {
+                path = path.add(elem);
+            }
+            this.basePath = path;
+        } else {
+            this.basePath = BlobPath.cleanPath();
+        }
+
+        int concurrentStreams = repositorySettings.settings().getAsInt("concurrent_streams", componentSettings.getAsInt("concurrent_streams", 5));
+        ExecutorService concurrentStreamPool = EsExecutors.newScaling(1, concurrentStreams, 5, TimeUnit.SECONDS, EsExecutors.daemonThreadFactory(settings, "[gcs_stream]"));
```

Comment:
This pull request is a kind of hacky because elasticsearch Snapshot API and Google's Storage API really don't match well. It explains why we had to use Piped streams which are not really efficient. Both APIs have been improved and we were able to start a freshly pull request for 2.x.

You can try and use this code, and yes you need shutdown the ExecutorService and await for its termination, that can takes lot of times in case of huge snapshots.